server/src/main/java/org/candlepin/model/CandlepinQuery.java

```diff
@@ -0,0 +1,254 @@
+     * with queries that can return extremely large data sets.
+     *
+     * @return
+     *  a list containing the results of executing this criteria
+     */
+    List<T> list();
+
+    /**
+     * Steps through the results of a column of the given query row-by-row, rather than dumping the
+     * entire query result into memory before processing it. This method will always pass the first
+     * column of each row to the processor.
+     *
+     * @param processor
+     *  A ResultProcessor instance to use for processing each result
+     *
+     * @return
+     *  the number of rows processed and sent to the result processor
+     */
+    int scroll(ResultProcessor<T> processor);
+
+    /**
+     * Steps through the results of a column of the given query row-by-row, rather than dumping the
+     * entire query result into memory before processing it.
+     *
+     * @param column
+     *  The zero-indexed offset of the column to process
+     *
+     * @param processor
+     *  A ResultProcessor instance to use for processing each result
+     *
+     * @return
+     *  the number of rows processed and sent to the result processor
+     */
+    int scroll(int column, ResultProcessor<T> processor);
+
+    /**
+     * Steps through the results of a query row-by-row, rather than dumping the entire query result
+     * into memory before processing it.
+     * <p/>
+     * If this method is called with eviction enabled, the result processor must manually persist
+     * and flush each object that is changed, as any unflushed changes will be lost when the object
+     * is evicted.
+     *
+     * @param evict
+     *  Whether or not to auto-evict queried objects after they've been processed
+     *
+     * @param processor
+     *  A ResultProcessor instance to use for processing each result
+     *
+     * @return
+     *  the number of rows processed and sent to the result processor
+     */
+    int scroll(int column, boolean evict, ResultProcessor<T> processor);
+
+    /**
+     * Steps through the results of a query row-by-row, rather than dumping the entire query result
+     * into memory before processing it. Unlike the base scroll method, this method sends each row
+     * to the processor without performing any preprocessing or cleanup.
+     *
+     * @param processor
+     *  A ResultProcessor instance to use for processing each row
+     *
+     * @return
+     *  the number of rows processed by the result processor
+     */
+    int scrollByRow(ResultProcessor<Object[]> processor);
+
+    /**
+     * Executes this query and iterates over the first column of the results. Other columns in
+     * each row are silently discarded.
+     * <p></p>
+     * WARNING: This method must be called from within a transaction, and the iterator must
+     * remain within the bounds of that transaction.
+     *
+     * @return
+     *  an iterator over the first column of the results
+     */
+    ResultIterator<T> iterate();
```

Comment:
The only reservation I have for keeping it around is that the return type is more explicit than the standard iterator() method returns. If we didn't need to worry about closing the iterator/cursor, it wouldn't be an issue.